src/charm.py

```diff
@@ -327,15 +327,16 @@ def _should_run_s3_migration(self, current_plan: Dict, s3info: Optional[S3Info])
             If no services are planned yet (first run) or S3 settings have changed.
         """
         # Properly type checks would require defining a complex TypedMap for the pebble plan
-        return not current_plan.services or (  # type: ignore
-            # Or S3 is enabled and one S3 parameter has changed
-            self.config.get("s3_enabled")
-            and s3info
-            and (
-                s3info.enabled != self.config.get("s3_enabled")
-                or s3info.region != self.config.get("s3_region")
-                or s3info.bucket != self.config.get("s3_bucket")
-                or s3info.endpoint != self.config.get("s3_endpoint")
+        return self.config.get("s3_enabled") and (
+            not current_plan.services  # type: ignore
```

Comment:
@benhoyt any thoughts on this? @arturo-seijas I'm guessing the comment on line 329 explains the reason for the type: ignore here?

Reply:
No, I don't think we should be `type: ignore`ing here. The type signature of `_should_run_s3_migration` is incorrect: `current_plan` should be a `pebble.Plan`, not a `Dict`. That gets rid of one mypy error. The other error you can fix by converting the result to a proper bool instead of an `and`ed truthy value, using `bool(result)`.

I've removed the `type: ignore` and made the following changes and it type checks nicely:

```diff
diff --git a/src/charm.py b/src/charm.py
index dec3ad0..c82417f 100755
--- a/src/charm.py
+++ b/src/charm.py
@@ -18,7 +18,7 @@ from ops.charm import ActionEvent, CharmBase, HookEvent
 from ops.framework import StoredState
 from ops.main import main
 from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
-from ops.pebble import ExecError
+from ops.pebble import ExecError, Plan
 
 logger = logging.getLogger(__name__)
 pgsql = ops.lib.use("pgsql", 1, "postgresql-charmers@lists.launchpad.net")
@@ -318,7 +318,7 @@ class DiscourseCharm(CharmBase):
         }
         return layer_config
 
-    def _should_run_s3_migration(self, current_plan: Dict, s3info: Optional[S3Info]) -> bool:
+    def _should_run_s3_migration(self, current_plan: Plan, s3info: Optional[S3Info]) -> bool:
         """Determine if the S3 migration is to be run.
 
         Args:
@@ -328,9 +328,8 @@ class DiscourseCharm(CharmBase):
         Returns:
             If no services are planned yet (first run) or S3 settings have changed.
         """
-        # Properly type checks would require defining a complex TypedMap for the pebble plan
-        return self.config.get("s3_enabled") and (
-            not current_plan.services  # type: ignore
+        result = self.config.get("s3_enabled") and (
+            not current_plan.services
             or (
                 s3info
                 and (
@@ -341,6 +340,7 @@ class DiscourseCharm(CharmBase):
                 )
             )
         )
+        return bool(result)
 
     def _are_db_relations_ready(self) -> bool:
         """Check if the needed database relations are established.
```

Reply:
Excellent, thank you @benhoyt!